### ISSUE TITLE
Annotation Processors Ignore TYPE_USE annotations on compiled classes

### ISSUE THREAD

**Issue by agentgt** [jspecify/jspecify#365]
TLDR; JSpecify might need more `java.lang.annotation.Target`s then just `TYPE_USE` for compatibility or some documented work around for annotation processors.

While playing with JSpecify and Eclipse null annotations which both are only `TYPE_USE` I found the rather disturbing possible bug with the JDK APT framework in that `TYPE_USE` annotations appear to be stripped from already compiled classes. I stress "appear" because the data is definitely in the class file it just the APT framework does not make it visible.

Let us say we have module `ModuleA` (aka jar artifact `ModuleA`) that has a class `ClassA`. `ClassA` has a nullable `TYPE_USE` annotations on some method return type we will call `ReturnTypeA`:

```java
public class ClassA {
 @Nullable ReturnTypeA someMethod() { return null; }
}
```

If we compile `ClassA` with an annotation processor enabled and somehow navigate to the method  `someMethod` and get the TypeMirror for `ReturnTypeA` and call `TypeMirror.getAnnotationMirrors()` the `@Nullable` will be there. Good. The type is preserved.

Now let us say we have `ModuleB` that has `ClassB` like:

```java
public class ClassB {
  public ClassA a;
}
```

And again if we compile `ModuleB` (and thus `ClassB`) with an annotation processor enabled and navigate to `ClassA` method  `someMethod` and get the TypeMirror for `ReturnTypeA` and call `TypeMirror.getAnnotationMirrors()` the `@Nullable` WILL BE MISSING regardless of Retention Policy.

This is pretty darn problematic for annotation processors like MapStruct (see https://github.com/mapstruct/mapstruct/issues/1243) or my own library JStachio where we want to know if the type allows nulls or not.

To be honest this seems like a JDK issue because almost all other information is available with TypeElement or TypeMirror as well as you can reflectively get access to the annotations with `java.lang.Method.getAnnotatedReturnType()`. Furthermore all other annotations that are NOT `TYPE_USE` are available. Perhaps this is one of the reasons Checkers target basically everything?

@kevinb9n have you guys ran across this at all?

**Comment by cpovirk** [comment]
Thanks very much for reporting. This is a case that we are sadly aware of but have been embarrassingly slow to do anything about.

The JDK bug for it is [JDK-8225377](https://bugs.openjdk.org/browse/JDK-8225377).

Many nullness checkers have grown their own workarounds (using internal JDK APIs), but we want to get it fixed, both so that those checkers can someday remove their workarounds (some of which are buggy, e.g., [ours](https://github.com/google/error-prone/blob/master/check_api/src/main/java/com/google/errorprone/dataflow/nullnesspropagation/NullnessAnnotations.java#L79-L80) :)) and so that annotation processors that just try to _use_ nullness information (like yours) can do so without needing to employ similar tricks.

**Comment by agentgt** [comment]
@cpovirk Thank you for the comment! It was extremely helpful. I tried ErrorPrones Symbol getRawAttributes hack and I think it could work but the required setup would be rather painful for consumers of the library.

For now I'm going to shelve our nullable support for [my library at the moment](https://github.com/jstachio/jstachio): https://github.com/jstachio/jstachio/issues/4.

Partly because of the hack but mostly as I'm not sure what level of support to provide as it is just a templating language.

As a side note I was wondering as a workaround to the jdk bug some JSpecify tool/implementation annotation processor could generate external annotations (such as eclipses external annotation format) as a class resource. Then downstream tools could check kind of like the non modular service loader (META-INF/services) if the annotations file is present.

Using my previous example `ModuleA` would be compiled with external annotation generator. `ModuleB` could then read the file (classpath resource). Not ideal but it is a tool even I could make ... I think.

**Comment by cpovirk** [comment]
You're welcome. I'm glad it was helpful. Thanks for the link to your project and for your interest.

As for the workaround: We've occasionally had conversations about developing a consensus format for external annotations. I don't know if we'd connected that with this current issue, though, with the idea of writing the information we already know into an external file so that it can be read without hacks.

To be fully general, we'd want to write that information for all classes on the classpath that might contain nullness annotations. And we'd want various tools to read it—which _might_ be straightforward if we can provide drop-in replacements for all the relevant `javax.lang` APIs. Overall, there's enough uncertainty (e.g., when a class in the dependencies is repackaged/shaded) that I think we'll probably still focus more on fixing javac—even though that, too, is uncertain :) But it's helpful to have another option to keep in mind and to inspire other ideas. And it probably _is_ a straightforward option for cases in which you care only about one or two tools and only about certain kinds of classes.

**Comment by rob-bygrave** [comment]
FYI: Just to say the JDK bug is fixed in 22 and back ported to 21.0.2, 17.0.11, 11.0.23:

[JDK-8319885](https://bugs.openjdk.org/browse/JDK-8319885)	21.0.2 	
[JDK-8321934](https://bugs.openjdk.org/browse/JDK-8321934)	17.0.11 	
[JDK-8322203](https://bugs.openjdk.org/browse/JDK-8322203) 11.0.23

**Comment by cushon** [comment]
Unfortunately the fix is no longer going to be in 21.0.2, and may be backed out of the upcoming 17u and 11u releases also: [JDK-8322883](https://bugs.openjdk.org/browse/JDK-8322883)

**Comment by SentryMan** [comment]
Just ran into this one today, it was quite baffling. I got around it by using a `TypeVisitor` on the typeMirror, so while I can't directly read `TypeMirror.getAnnotationMirrors()` I can use the visitor to [extract the annotations](https://github.com/avaje/avaje-record-builder/blob/8f430a1324bbb311f7ccd6cdcee388a84505d396/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Utils.java#L35) and do the checks.

**Comment by cushon** [comment]
@wmdietl the `javac fails with "Unable to implement <class> method"` issue was fixed by https://github.com/micronaut-projects/micronaut-core/pull/10293. There are probably other similar issues in the wild, so the question is how many processors are affected and what the impact will be. I wrote up a CSR to discuss the compatibility impact ([JDK-8323093](https://bugs.openjdk.org/browse/JDK-8323093)). If it does get backed out for now, I think we could try again for the next update release in July.

**Comment by sdeleuze** [comment]
As I start the migration of Spring Framework `main` branch (future Spring Framework 7.0) that will very significantly increase JSpecify reach, I would welcome an update on the backport of the JDK fix to 17 and 21 (we don't care about Java 11 as Spring baseline is Java 17). Spring contains runtime checks of null-safety annotation so it is very important that Spring Boot applications works as expected once migrated to JSpecify as of Java 17.

**Comment by cushon** [comment]
> I would welcome an update on the backport of the JDK fix to 17 and 21

Re-doing the backport is tracked by [JDK-8341779](https://bugs.openjdk.org/browse/JDK-8341779). There was a compiler-dev@ thread from July/August linked from the bug, I don't have more recent updates.

**Comment by agentgt** [comment]
@sdeleuze I'm not sure Spring has to worry about this as it is mostly reflection based.

That is the JSpecify Nullable annotations are visible with reflection. 

Is Spring using some annotation processors for tooling or code generation (that I'm unaware of)?

**Comment by sdeleuze** [comment]
@agentgt Spring Boot is using 2 annotation processors `org.springframework.boot.autoconfigureprocessor.AutoConfigureAnnotationProcessor` and `org.springframework.boot.configurationprocessor.ConfigurationMetadataAnnotationProcessor`. In `org.springframework.boot.configurationprocessor.MetadataGenerationEnvironment#hasNullableAnnotation`, we check for the existence of `org.springframework.lang.Nullable`, and we need as part of Spring Boot 4 to add similar support for `org.jspecify.annotations.Nullable`. If I am not mistaken, for now it is only possible with with Java 22+, so it would be a significant regressions for Java 17 and 21 users.

@cushon @kevinb9n Looking forward to synchronize on how we can collaborate for a new pushing for backporting this fix to Java 17 and 21.

**Comment by CfFanDuel** [comment]
Is this fixed then?

https://bugs.openjdk.org/browse/JDK-8341779

suggests the backport might have taken place but I'm not very familiar with looking through jdk tickets.

**Comment by cpovirk** [comment]
I think we're waiting for 21.0.8 to be released. That looks to be targeted for [July 15](https://wiki.openjdk.org/display/JDKUpdates/JDK+21u#:~:text=OpenJDK%2021.0.8%20released). (The section title there says "OpenJDK 21.0.*9*," but it looks pretty clearly like a typo.)

**Comment by cpovirk** [comment]
(You can also still already get the fix by using a javac from JDK 22 or higher, passing flags like `--release 21` to maintain compatibility.)

**Comment by agentgt** [comment]
I forgot to mention another JDK bug that is relevant to this one.

https://bugs.openjdk.org/browse/JDK-8281238

I'm remiss that I did not mention it sooner so apologies on the late notice. The TLDR is prior to JDK 19 `toString` on a TypeMirror produced the incorrect type signature if TYPE_USE annotations are present.

I have known about the bug for some time but forgot about because I had a custom Visitor to generate the correct full qualified type.

However note that plenty of folks still just `toString()` a TypeMirror instead of using a Visitor.

So here is the kicker once the backport ("type annotations are not visible to javac plugins across compilation boundaries") is fixed and it is applied to anything earlier than 19 there will likely be issues for any processors that blindly `toString()` TypeMirrors. This is because the backport will make the possibility of having a TYPE_USE in the signature of anything far greater.

Luckily this toString issue was fixed in 19 so 21 backport will be fine. However if they backport to 17 be aware.

Pinging @sdeleuze incase they are still targeting 17 where I think this would be the case.

(I also have a visitor laying around that will properly toString a TypeMirror regardless of version that I can add later. It can also remove the TYPE_USEs annotations).

**Comment by sdeleuze** [comment]
@cushon Any thoughts on backporting https://bugs.openjdk.org/browse/JDK-8281238 to Java 17 as well when the backport will be applied to Java 17?

**Comment by cushon** [comment]
I'll make a note to consider JDK-8284220/JDK-8342934 if the fix for JDK-8225377 is backported to 17
